Cheque a versão do sistema operacional e se o cuda está presente no diretório

O colab deve estar em modo GPU

É importante que tenha a pasta do projeto C3D disponível em: [link text](https://github.com/devjaynemorais/C3D/) contida no seu Drive pessoal.

In [0]:
# check Utubtu version
!lsb_release -a
# by default, CUDA is enabled under GPU mode of Google Colab
!ls /usr/local

No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.3 LTS
Release:	18.04
Codename:	bionic
bin   cuda-10.0  etc	include  LICENSE.txt  sbin	 share	xgboost
cuda  cuda-10.1  games	lib	 man	      setup.cfg  src


A saída esperada deve ser semelhante a:

`No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.3 LTS
Release:	18.04
Codename:	bionic
bin   cuda-10.0  etc	include  LICENSE.txt  sbin	 share	xgboost
cuda  cuda-10.1  games	lib	 man	      setup.cfg  src`

In [0]:
# optional
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


# Note


```
As of the CUDA 8 release, gcc 5.3 is fully supported on Ubuntu 16.06 and Fedora 23.
As of the CUDA 9 release, gcc 6 is fully supported on Ubuntu 16.04, Ubuntu 17.04 and Fedora 25.
The CUDA 9.2 release adds support for gcc 7
```



In [0]:
# make g++, nvcc and boost work together (compatible). 
# Better install gcc before anything else
# update-alternatives --install [link] [name] [path] [priority]

!sudo apt update
!sudo apt install g++-5
!sudo update-alternatives --remove-all gcc 
!sudo update-alternatives --remove-all g++

!sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-5 20
!sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-5 20

!sudo update-alternatives --install /usr/bin/cc cc /usr/bin/gcc 30
!sudo update-alternatives --set cc /usr/bin/gcc

!sudo update-alternatives --install /usr/bin/c++ c++ /usr/bin/g++ 30
!sudo update-alternatives --set c++ /usr/bin/g++

Hit:1 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease
Get:2 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Hit:3 http://archive.ubuntu.com/ubuntu bionic InRelease
Hit:4 http://ppa.launchpad.net/marutter/c2d4u3.5/ubuntu bionic InRelease
Get:5 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Hit:6 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran35/ InRelease
Get:7 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]
Ign:8 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Ign:9 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Hit:10 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release
Hit:11 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Fetched 252 kB in 1s (355 kB/s)
Reading package lists... Done
Building dependency tree       
Reading s

In [0]:
!sudo apt-get install -f

# install dependencies

!sudo apt-get install -y libprotobuf-dev libleveldb-dev libsnappy-dev libhdf5-serial-dev protobuf-compiler
!sudo apt-get install libatlas-base-dev
!sudo apt-get install -y --no-install-recommends libboost-all-dev
!sudo apt-get install -y libgflags-dev libgoogle-glog-dev liblmdb-dev

Reading package lists... Done
Building dependency tree       
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 91 not upgraded.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libleveldb-dev is already the newest version (1.20-2).
libprotobuf-dev is already the newest version (3.0.0-9.1ubuntu1).
libsnappy-dev is already the newest version (1.1.7-1).
libhdf5-serial-dev is already the newest version (1.10.0-patch1+docs-4).
protobuf-compiler is already the newest version (3.0.0-9.1ubuntu1).
0 upgraded, 0 newly installed, 0 to remove and 91 not upgraded.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libatlas-base-dev is already the newest version (3.10.3-5).
0 upgraded, 0 newly installed, 0 to remove and 91 not upgraded.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libboost-all-dev is already the newest version 

# Note
```
CUDA, OPENCV
• BLAS (Basic Linear Algebra Subprograms): operations like matrix
multiplication, matrix addition, both implementation for CPU(cBLAS) and
GPU(cuBLAS). provided by MKL(INTEL), ATLAS, openBLAS, etc.
• Boost: a c++ library. > Use some of its math functions and shared_pointer.
• glog,gflags provide logging & command line utilities. > Essential for
debugging.
• leveldb, lmdb: database io for your program. > Need to know this for
preparing your own data.
• protobuf: an efficient and flexible way to define data structure. > Need to
know this for defining new layer
```



Faça o clone da seguinte repositório. Atenção, para manter essa versão do notebook do Colab estável foi criado o branch `C3DVF-Colab`, em caso de alteração após 19/11/2019 no branch `master`,

In [0]:
%cd /content/gdrive/My\ Drive/
!git clone https://github.com/devjaynemorais/C3D.git

/content/gdrive/My Drive
Cloning into 'C3D'...
remote: Enumerating objects: 31, done.
remote: Counting objects: 100% (31/31), done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 1782 (delta 16), reused 0 (delta 0), pack-reused 1751
Receiving objects: 100% (1782/1782), 38.51 MiB | 15.61 MiB/s, done.
Resolving deltas: 100% (620/620), done.
Checking out files: 100% (1256/1256), done.


In [0]:
%cd /content/gdrive/My\ Drive/C3D/C3D-v1.0/
!ls

/content/gdrive/My Drive/C3D/C3D-v1.0
caffe.cloc	 docs	   INSTALL.md	    Makefile.config.example  scripts
CONTRIBUTORS.md  examples  Makefile	    matlab		     src
data		 include   Makefile.config  python		     tools


Saída esperada:

`/content/gdrive/My Drive/C3D/C3D-v1.0
caffe.cloc	 examples    Makefile.config			      python
CONTRIBUTORS.md  include     Makefile.config.example		      scripts
data		 INSTALL.md  matlab				      src
docs		 Makefile    output_novelty_features_c3d_extract.zip  tools`

Atualizando as bibliotecas de vínculo dinâmico


In [0]:
!sudo ldconfig /usr/local/cuda/lib64

/sbin/ldconfig.real: /usr/local/lib/python3.6/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link



A saída esperada deve ser:

`/sbin/ldconfig.real: /usr/local/lib/python3.6/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link`

In [0]:
!sudo apt-get install nvidia-modprobe


Reading package lists... Done
Building dependency tree       
Reading state information... Done
nvidia-modprobe is already the newest version (440.64.00-0ubuntu1).
0 upgraded, 0 newly installed, 0 to remove and 91 not upgraded.


In [0]:
!pip show protobuf

Name: protobuf
Version: 3.10.0
Summary: Protocol Buffers
Home-page: https://developers.google.com/protocol-buffers/
Author: None
Author-email: None
License: 3-Clause BSD License
Location: /usr/local/lib/python3.6/dist-packages
Requires: six, setuptools
Required-by: tensorflow, tensorflow-metadata, tensorflow-hub, tensorflow-datasets, tensorboard, googleapis-common-protos, google-cloud-bigquery, google-api-core, chainer


Observe o modelo da Placa de vídeo utilizada para alterar a variável `CUDA_ARCH` no arquivo `C3D/C3D-v1.0/Makefile.config`. Em caso de dúvidas sobre qual parâmetro utilizar para a variável, consulte: [Cuda-GPUs](https://developer.nvidia.com/cuda-gpus).

Para este notebook, a placa de vídeo é a Tesla K80. Portanto, adicionaremos a linha `-gencode arch=compute_37,code=sm_37 \` na variável `CUDA_ARCH` presente no arquivo.



In [0]:
!nvidia-smi

Sun Apr 26 22:31:43 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64.00    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   37C    P8     7W /  75W |      0MiB /  7611MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU  

A saída esperada deve ser semelhante a:

`Sun Apr 26 22:31:43 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64.00    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   37C    P8     7W /  75W |      0MiB /  7611MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+`

In [0]:
!cat /usr/local/cuda/version.txt

CUDA Version 10.1.243


Vá até o arquivo `/content/gdrive/My\ Drive/C3D/C3D-v1.0/Makefile.config` no Drive e abra-o com um editor de texto e adicione a linha correspondente a variável do `CUDA_ARCH` de acordo com a versão correta da laca de vídeo disponibilizada pelo Google Colaboratory: Ex.: `-gencode arch=compute_61,code=sm_61` na variável `CUDA_ARCH`.

[Como neste exemplo para uma Testa P4](https://drive.google.com/open?id=1XO-FY0S3WW3QMIR7n9V403oRUe_uOB3h)

Confira se o Makefile.config mudou!

**Obs.:** Às vezes o arquivo editado demora a ser atualizado aqui no notebook, sugiro atualizar a página e apenas prosseguir quando o arquivo exibir a atualização realizada.



In [17]:
!cat /content/gdrive/My\ Drive/C3D/C3D-v1.0/Makefile.config

## Refer to http://caffe.berkeleyvision.org/installation.html
# Contributions simplifying and improving our build system are welcome!
# C3D_V1.0

# CUDA directory contains bin/ and lib/ directories that we need.
CUDA_DIR := /usr/local/cuda

# CUDA architecture setting: going with all of them (up to CUDA 5.5 compatible).
# For the latest architecture, you need to install CUDA >= 6.0 and uncomment
# the *_50 lines below.
# DoLongBien, CUDA 9.2 ERROR nvcc fatal   : Unsupported gpu architecture 'compute_20'
# CUDA architecture setting: going with all of them.
# For CUDA < 6.0, comment the *_50 through *_61 lines for compatibility.
# For CUDA < 8.0, comment the *_60 and *_61 lines for compatibility.
# For CUDA >= 9.0, comment the *_20 and *_21 lines for compatibility.
CUDA_ARCH := 	-gencode arch=compute_30,code=sm_30 \
		-gencode arch=compute_35,code=sm_35 \
		-gencode=arch=compute_50,code=sm_50 \
		-gencode=arch=compute_61,code=compute_61

# Uncomment if your OpenCV version >= 3.0
 OPENCV_

## Compilação

In [18]:
# compile

%cd /content/gdrive/My\ Drive/C3D/C3D-v1.0/
!make clean
!make all -j$(nproc)

/content/gdrive/My Drive/C3D/C3D-v1.0
protoc --proto_path=src --cpp_out=build/src src/caffe/proto/caffe.proto
protoc --proto_path=src --cpp_out=build/src src/caffe/proto/caffe_pretty_print.proto
[libprotobuf WARNING google/protobuf/compiler/parser.cc:547] No syntax specified for the proto file: caffe/proto/caffe.proto. Please use 'syntax = "proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted to proto2 syntax.)
[libprotobuf WARNING google/protobuf/compiler/parser.cc:547] No syntax specified for the proto file: caffe/proto/caffe_pretty_print.proto. Please use 'syntax = "proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted to proto2 syntax.)
[libprotobuf WARNING google/protobuf/compiler/parser.cc:547] No syntax specified for the proto file: caffe/proto/caffe.proto. Please use 'syntax = "proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted to proto2 syntax.)

/usr/local/cuda/bin/nvcc -ccbin=/usr/bin/g++ -Xcompiler -fPIC -DNDEB

Caso seja apresentadas algumas linhas como a linha abaixo:

`#warning "math_functions.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^`

Não há problemas!

As últimas linhas da saída esperada devem ser semelhantes a:

`/usr/bin/g++ build/examples/cifar10/convert_cifar_data.o build/lib/libcaffe.a -o build/examples/cifar10/convert_cifar_data.bin -pthread -fPIC -DNDEBUG -O2 -I/usr/local/include/python2.7 -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/include -I/usr/include/hdf5/serial/ -Ibuild/src -I./src -I./include -I/usr/local/cuda/include -L/usr/local/lib -L/usr/local/lib -L/usr/lib -L/usr/lib/x86_64-linux-gnu/hdf5/serial -L/usr/local/cuda/lib64 -L/usr/local/cuda/lib -lcudart -lcublas -lcurand -lpthread -lglog -lprotobuf -lleveldb -lsnappy -lboost_system -lhdf5_hl -lhdf5 -lopencv_core -lopencv_highgui -lopencv_imgproc -lopencv_imgcodecs -lopencv_videoio -lcblas -latlas -Wall`

`/usr/bin/g++ build/examples/mnist/convert_mnist_data.o build/lib/libcaffe.a -o build/examples/mnist/convert_mnist_data.bin -pthread -fPIC -DNDEBUG -O2 -I/usr/local/include/python2.7 -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/include -I/usr/include/hdf5/serial/ -Ibuild/src -I./src -I./include -I/usr/local/cuda/include -L/usr/local/lib -L/usr/local/lib -L/usr/lib -L/usr/lib/x86_64-linux-gnu/hdf5/serial -L/usr/local/cuda/lib64 -L/usr/local/cuda/lib -lcudart -lcublas -lcurand -lpthread -lglog -lprotobuf -lleveldb -lsnappy -lboost_system -lhdf5_hl -lhdf5 -lopencv_core -lopencv_highgui -lopencv_imgproc -lopencv_imgcodecs -lopencv_videoio -lcblas -latlas -Wall
`



In [19]:
!make test

/usr/bin/g++ src/caffe/test/test_benchmark.cpp -pthread -fPIC -DNDEBUG -O2 -I/usr/local/include/python2.7 -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/include -I/usr/include/hdf5/serial/ -Ibuild/src -I./src -I./include -I/usr/local/cuda/include -c -o build/src/caffe/test/test_benchmark.o

/usr/bin/g++ src/caffe/test/test_blob.cpp -pthread -fPIC -DNDEBUG -O2 -I/usr/local/include/python2.7 -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/include -I/usr/include/hdf5/serial/ -Ibuild/src -I./src -I./include -I/usr/local/cuda/include -c -o build/src/caffe/test/test_blob.o

/usr/bin/g++ src/caffe/test/test_common.cpp -pthread -fPIC -DNDEBUG -O2 -I/usr/local/include/python2.7 -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/include -I/usr/include/hdf5/serial/ -Ibuild/src -I./src -I./include -I/usr/local/cuda/include -c -o build/src/caffe/test/test_common.o

/usr/bin/g++ src/caffe/test/test_concat_layer.cpp -pthread -fP

In [0]:
# Opcional

#%cd /content/gdrive/My\ Drive/C3D/C3D-v1.0/
#!make runtest

In [21]:
!cat /proc/driver/nvidia/version

NVRM version: NVIDIA UNIX x86_64 Kernel Module  418.67  Sat Apr  6 03:07:24 CDT 2019
GCC version:  Selected multilib: .;@m64


### Vamos baixar os Pesos

In [22]:
%cd /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/
!wget -O conv3d_deepnetA_sport1m_iter_1900000 https://www.dropbox.com/s/vr8ckp0pxgbldhs/conv3d_deepnetA_sport1m_iter_1900000?dl=0

/content/gdrive/My Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction
--2020-04-26 22:49:25--  https://www.dropbox.com/s/vr8ckp0pxgbldhs/conv3d_deepnetA_sport1m_iter_1900000?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.65.1, 2620:100:6021:1::a27d:4101
Connecting to www.dropbox.com (www.dropbox.com)|162.125.65.1|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/vr8ckp0pxgbldhs/conv3d_deepnetA_sport1m_iter_1900000 [following]
--2020-04-26 22:49:25--  https://www.dropbox.com/s/raw/vr8ckp0pxgbldhs/conv3d_deepnetA_sport1m_iter_1900000
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc3daa118881baacd3a2d0330ca3.dl.dropboxusercontent.com/cd/0/inline/A2mMYcwgifszgwaS5bVcyM8JhBhpdzOpTbxPIiJFOY6q-a-dxDcUEP7xR8-NQCSLKcj2RtnaWFPHATdvnxNGeObYow1dFeWGdgeZa8YwgVaeKA/file# [following]
--2020-04-26 22:49:26--  https://uc3daa118881baacd3a2d0330ca3.dl.dropboxusercontent.c

Caso dê certo, as últimas linhas da saída esperada devem ser semelhantes a:

`HTTP request sent, awaiting response... 200 OK`
`Length: 319966293 (305M) [application/octet-stream]`
`Saving to: ‘conv3d_deepnetA_sport1m_iter_1900000’`

`conv3d_deepnetA_spo 100%[===================>] 305.14M  45.7MB/s    in 6.8s` 

`2020-04-26 16:55:31 (44.8 MB/s) - ‘conv3d_deepnetA_sport1m_iter_1900000’ saved [319966293/319966293]`

# Extract feature, check output folder in c3d_sport1m_feature_extraction_video.sh 

Teste de extração de características C3D com os vídeos e arquivos in/output usados de exemplo e já disponíveis.




In [23]:
# extract feature
%cd /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction
!bash c3d_sport1m_feature_extraction_video.sh

/content/gdrive/My Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction
E0426 22:49:55.788264 11631 extract_image_features.cpp:64] Using GPU #0
I0426 22:49:55.791014 11631 net.cpp:76] Creating Layer data
I0426 22:49:55.791036 11631 net.cpp:112] data -> data
I0426 22:49:55.791049 11631 net.cpp:112] data -> label
I0426 22:49:55.791213 11631 video_data_layer.cpp:283] Opening file prototxt/input_list_video.txt
I0426 22:49:55.791877 11631 video_data_layer.cpp:317] A total of 16 video chunks.
I0426 22:49:55.945551 11631 video_data_layer.cpp:365] output data size: 50,3,16,112,112
I0426 22:49:55.945580 11631 video_data_layer.cpp:387] Loading mean file from sport1m_train16_128_mean.binaryproto
I0426 22:49:56.019903 11631 net.cpp:127] Top shape: 50 3 16 112 112 (30105600)
I0426 22:49:56.019942 11631 net.cpp:127] Top shape: 50 1 1 1 1 (50)
I0426 22:49:56.019950 11631 net.cpp:159] data does not need backward computation.
I0426 22:49:56.019994 11631 net.cpp:76] Creating Layer conv1a
I0426 22:49:56.02

Caso dê certo, as últimas linhas da saída esperada são:

`I0426 16:55:58.611307  5215 net.cpp:177] Memory required for Data 5653289408`

`E0426 16:55:59.296900  5215 extract_image_features.cpp:81] Extracting features for 1 batches`

`E0426 16:56:02.655942  5215 extract_image_features.cpp:117] Extracted features of 16 images.`

`E0426 16:56:02.655977  5215 extract_image_features.cpp:121] Successfully extracted 16 features!`

Visualizar os arquivos da camada fc6-1, fc7-1 e prob extraídos a cada 16 frames de cada vídeo.

In [24]:
!ls /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/output/c3d/v_ApplyEyeMakeup_g01_c01/

000000.fc6-1  000032.fc6-1  000064.fc6-1  000096.fc6-1	000128.fc6-1
000000.fc7-1  000032.fc7-1  000064.fc7-1  000096.fc7-1	000128.fc7-1
000000.prob   000032.prob   000064.prob   000096.prob	000128.prob
000016.fc6-1  000048.fc6-1  000080.fc6-1  000112.fc6-1	000144.fc6-1
000016.fc7-1  000048.fc7-1  000080.fc7-1  000112.fc7-1	000144.fc7-1
000016.prob   000048.prob   000080.prob   000112.prob	000144.prob


Caso dê certo, a saída esperada é:

`000000.fc6-1  000032.fc6-1  000064.fc6-1  000096.fc6-1	000128.fc6-1
000000.fc7-1  000032.fc7-1  000064.fc7-1  000096.fc7-1	000128.fc7-1
000000.prob   000032.prob   000064.prob   000096.prob	000128.prob
000016.fc6-1  000048.fc6-1  000080.fc6-1  000112.fc6-1	000144.fc6-1
000016.fc7-1  000048.fc7-1  000080.fc7-1  000112.fc7-1	000144.fc7-1
000016.prob   000048.prob   000080.prob   000112.prob	000144.prob`


# Gerando arquivos in/output para nosso próprio conjunto de dados.

Primeiro, remova o arquivo de script existente, casos exista.

In [25]:
%cd /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/
!rm -rf /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/generate_inout_file.py

/content/gdrive/My Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction


## Upload do dataset

Vá ao drive e faça upload dos vídeos na pasta `/content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/`

Confira se os vídeos estão na pasta indicada:

In [26]:
!ls /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/

avi			v_JumpRope_g22_c01.avi	 v_SalsaSpin_g04_c05.avi
frm			v_JumpRope_g23_c05.avi	 v_SalsaSpin_g10_c05.avi
generate_inout_file.py	v_Punch_g19_c03.avi	 v_SalsaSpin_g18_c02.avi
v_JumpRope_g09_c06.avi	v_Punch_g21_c05.avi
v_JumpRope_g16_c02.avi	v_SalsaSpin_g03_c06.avi


Mais exemplos

In [0]:
#!\cp /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/Assault004_x264.mp4 /content/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/Assault004_x264.mp4
#!\cp /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/Explosion008_x264.mp4 /content/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/Explosion008_x264.mp4
#!\cp /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/Explosion009_x264.mp4 /content/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/Explosion009_x264.mp4
#!\cp /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/Normal_Videos_006_x264.mp4 /content/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/Normal_Videos_006_x264.mp4

## Vamos gerar o arquivo com a lista dos inputs


In [28]:
!cat /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/generate_inout_file.py

# Copy of https://github.com/dolongbien/HumanBehaviorBKU/blob/master/C3D/generate_inout_file.py

#cd /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/Traning_Normal_Part1_Video
import os, os.path

# import the necessary packages
#from ..convenience import is_cv3

import cv2
import glob
import math

f_input = open('prototxt/input_list_video.txt', 'w')
f_output = open('prototxt/output_list_video_prefix.txt','w')
sh_out = open('c3d_sport1m_feature_extraction_video.sh', 'w')

interval_v = 16

cmd1 = 'time GLOG_logtosterr=1 ../../build/tools/extract_image_features.bin prototxt/c3d_sport1m_feature_extractor_video.prototxt conv3d_deepnetA_sport1m_iter_1900000'

cmd2 = 'prototxt/output_list_video_prefix.txt fc6-1'

gpu_id = 0
batch_size = 50
batch_num = 1
max_batch_num = 1
result = 0

# read file path name in alphabet order
file_list_v = sorted(glob.glob1("/content/gdrive/My Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/input", "*.avi"))

print('files.....')
pr

In [0]:
#! rm -rf /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/Assault004_x264.mp4

In [0]:
#!ls -la /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/generate_inout_file.py

Gere os arquivos in/out para o modelo C3D

In [31]:
!python /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/input/generate_inout_file.py

files.....
['v_JumpRope_g09_c06.avi', 'v_JumpRope_g16_c02.avi', 'v_JumpRope_g22_c01.avi', 'v_JumpRope_g23_c05.avi', 'v_Punch_g19_c03.avi', 'v_Punch_g21_c05.avi', 'v_SalsaSpin_g03_c06.avi', 'v_SalsaSpin_g04_c05.avi', 'v_SalsaSpin_g10_c05.avi', 'v_SalsaSpin_g18_c02.avi']
1	frames=491
IF -  10
result  10
2	frames=486
IF -  10
result  20
3	frames=198
ELSE -  10
result  24
4	frames=242
ELSE -  10
result  29
5	frames=300
ELSE -  10
result  35
6	frames=300
ELSE -  10
result  41
7	frames=168
ELSE -  10
result  45
8	frames=333
ELSE -  10
result  52
9	frames=168
ELSE -  10
result  56
10	frames=168
ELSE -  10
result  60
-------- PARAMETER --------
GPU ID is 0
Batch size is 50
Number of batch is 60


In [32]:
!cat /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/prototxt/input_list_video.txt

input/v_JumpRope_g09_c06.avi 0 0
input/v_JumpRope_g09_c06.avi 16 0
input/v_JumpRope_g09_c06.avi 32 0
input/v_JumpRope_g09_c06.avi 48 0
input/v_JumpRope_g09_c06.avi 64 0
input/v_JumpRope_g09_c06.avi 80 0
input/v_JumpRope_g09_c06.avi 96 0
input/v_JumpRope_g09_c06.avi 112 0
input/v_JumpRope_g09_c06.avi 128 0
input/v_JumpRope_g09_c06.avi 144 0
input/v_JumpRope_g09_c06.avi 160 0
input/v_JumpRope_g09_c06.avi 176 0
input/v_JumpRope_g09_c06.avi 192 0
input/v_JumpRope_g09_c06.avi 208 0
input/v_JumpRope_g09_c06.avi 224 0
input/v_JumpRope_g09_c06.avi 240 0
input/v_JumpRope_g09_c06.avi 256 0
input/v_JumpRope_g09_c06.avi 272 0
input/v_JumpRope_g09_c06.avi 288 0
input/v_JumpRope_g09_c06.avi 304 0
input/v_JumpRope_g09_c06.avi 320 0
input/v_JumpRope_g09_c06.avi 336 0
input/v_JumpRope_g09_c06.avi 352 0
input/v_JumpRope_g09_c06.avi 368 0
input/v_JumpRope_g09_c06.avi 384 0
input/v_JumpRope_g09_c06.avi 400 0
input/v_JumpRope_g09_c06.avi 416 0
input/v_JumpRope_g09_c06.avi 432 0
input/v_JumpRope_g09_c06.avi

In [33]:
!cat /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction/prototxt/output_list_video_prefix.txt

output/c3d/v_JumpRope_g09_c06/000000
output/c3d/v_JumpRope_g09_c06/000016
output/c3d/v_JumpRope_g09_c06/000032
output/c3d/v_JumpRope_g09_c06/000048
output/c3d/v_JumpRope_g09_c06/000064
output/c3d/v_JumpRope_g09_c06/000080
output/c3d/v_JumpRope_g09_c06/000096
output/c3d/v_JumpRope_g09_c06/000112
output/c3d/v_JumpRope_g09_c06/000128
output/c3d/v_JumpRope_g09_c06/000144
output/c3d/v_JumpRope_g09_c06/000160
output/c3d/v_JumpRope_g09_c06/000176
output/c3d/v_JumpRope_g09_c06/000192
output/c3d/v_JumpRope_g09_c06/000208
output/c3d/v_JumpRope_g09_c06/000224
output/c3d/v_JumpRope_g09_c06/000240
output/c3d/v_JumpRope_g09_c06/000256
output/c3d/v_JumpRope_g09_c06/000272
output/c3d/v_JumpRope_g09_c06/000288
output/c3d/v_JumpRope_g09_c06/000304
output/c3d/v_JumpRope_g09_c06/000320
output/c3d/v_JumpRope_g09_c06/000336
output/c3d/v_JumpRope_g09_c06/000352
output/c3d/v_JumpRope_g09_c06/000368
output/c3d/v_JumpRope_g09_c06/000384
output/c3d/v_JumpRope_g09_c06/000400
output/c3d/v_JumpRope_g09_c06/000416
o

In [34]:
!cat c3d_sport1m_feature_extraction_video.sh

rm -rf output/c3d/v_JumpRope_g09_c06
mkdir -p output/c3d/v_JumpRope_g09_c06
rm -rf output/c3d/v_JumpRope_g16_c02
mkdir -p output/c3d/v_JumpRope_g16_c02
rm -rf output/c3d/v_JumpRope_g22_c01
mkdir -p output/c3d/v_JumpRope_g22_c01
rm -rf output/c3d/v_JumpRope_g23_c05
mkdir -p output/c3d/v_JumpRope_g23_c05
rm -rf output/c3d/v_Punch_g19_c03
mkdir -p output/c3d/v_Punch_g19_c03
rm -rf output/c3d/v_Punch_g21_c05
mkdir -p output/c3d/v_Punch_g21_c05
rm -rf output/c3d/v_SalsaSpin_g03_c06
mkdir -p output/c3d/v_SalsaSpin_g03_c06
rm -rf output/c3d/v_SalsaSpin_g04_c05
mkdir -p output/c3d/v_SalsaSpin_g04_c05
rm -rf output/c3d/v_SalsaSpin_g10_c05
mkdir -p output/c3d/v_SalsaSpin_g10_c05
rm -rf output/c3d/v_SalsaSpin_g18_c02
mkdir -p output/c3d/v_SalsaSpin_g18_c02
time GLOG_logtosterr=1 ../../build/tools/extract_image_features.bin prototxt/c3d_sport1m_feature_extractor_video.prototxt conv3d_deepnetA_sport1m_iter_1900000 0 50 60 prototxt/output_list_video_prefix.txt fc6-1

## Extract Features

In [35]:
# extract feature
%cd /content/gdrive/My\ Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction
!bash c3d_sport1m_feature_extraction_video.sh

/content/gdrive/My Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction
E0426 22:53:22.198750 11945 extract_image_features.cpp:64] Using GPU #0
I0426 22:53:22.201416 11945 net.cpp:76] Creating Layer data
I0426 22:53:22.201437 11945 net.cpp:112] data -> data
I0426 22:53:22.201505 11945 net.cpp:112] data -> label
I0426 22:53:22.201704 11945 video_data_layer.cpp:283] Opening file prototxt/input_list_video.txt
I0426 22:53:22.202499 11945 video_data_layer.cpp:317] A total of 173 video chunks.
I0426 22:53:22.357290 11945 video_data_layer.cpp:365] output data size: 50,3,16,112,112
I0426 22:53:22.357321 11945 video_data_layer.cpp:387] Loading mean file from sport1m_train16_128_mean.binaryproto
I0426 22:53:22.431941 11945 net.cpp:127] Top shape: 50 3 16 112 112 (30105600)
I0426 22:53:22.432003 11945 net.cpp:127] Top shape: 50 1 1 1 1 (50)
I0426 22:53:22.432013 11945 net.cpp:159] data does not need backward computation.
I0426 22:53:22.432049 11945 net.cpp:76] Creating Layer conv1a
I0426 22:53:22.4

A saída deve ser semelhante as linhas abaixo:

`I0426 21:00:39.326395  5841 net.cpp:177] Memory required for Data 5653289408`

`E0426 21:00:40.377143  5841 extract_image_features.cpp:81] Extracting features for 60 batches`

`E0426 21:00:43.330047  5841 extract_image_features.cpp:117] Extracted features of 50 images.`

`E0426 21:00:58.617430  5841 extract_image_features.cpp:121] Successfully extracted 173 features!`

`real	0m24.586s`

`user	0m23.430s`

`sys	0m5.933s`

In [36]:
!ls output/c3d

v_ApplyEyeMakeup_g01_c01  v_JumpRope_g22_c01  v_SalsaSpin_g03_c06
v_BaseballPitch_g01_c01   v_JumpRope_g23_c05  v_SalsaSpin_g04_c05
v_JumpRope_g09_c06	  v_Punch_g19_c03     v_SalsaSpin_g10_c05
v_JumpRope_g16_c02	  v_Punch_g21_c05     v_SalsaSpin_g18_c02


Escolha um diretório correspondente a um vídeo e confira se os arquivos foram gerados

In [37]:
!ls -la output/c3d/v_JumpRope_g09_c06	  

total 495
-rw------- 1 root root 16404 Apr 26 22:53 000000.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000016.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000032.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000048.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000064.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000080.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000096.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000112.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000128.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000144.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000160.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000176.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000192.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000208.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000224.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000240.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000256.fc6-1
-rw------- 1 root root 16404 Apr 26 22:53 000272.fc6-1


Features Extraídas! :D

## Zip

In [38]:
!zip -r output_novelty_features_c3d_extract.zip output/c3d/

  adding: output/c3d/ (stored 0%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/ (stored 0%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000000.fc7-1 (deflated 80%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000016.fc7-1 (deflated 79%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000032.fc7-1 (deflated 78%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000048.fc7-1 (deflated 77%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000064.fc7-1 (deflated 79%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000080.fc7-1 (deflated 79%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000096.fc7-1 (deflated 83%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000112.fc7-1 (deflated 80%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000128.fc7-1 (deflated 79%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000144.fc7-1 (deflated 81%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000000.fc6-1 (deflated 56%)
  adding: output/c3d/v_ApplyEyeMakeup_g01_c01/000016.fc6-1 (deflated 56%)
  adding: output/c3

## Download

In [39]:
!pwd

/content/gdrive/My Drive/C3D/C3D-v1.0/examples/c3d_feature_extraction


In [0]:
from google.colab import files
files.download('output_novelty_features_c3d_extract.zip') 

Faça download do arquivo zipado com as features e bom trabalho!